### PR TITLE
[onnx] Add `onnx-to-torch` lowering for random ops

### PR DESCRIPTION
**Pull request by vivekkhandelwal1**

### CODE REVIEW

**Review by renxida — APPROVED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 1]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+          return failure();
+        }
+
+        dtypeIntTorch = onnxDtypeIntToTorchDtypeInt(dtypeIntOnnx);
+        if (dtypeIntTorch == -1) {
+          return rewriter.notifyMatchFailure(
+              binder.op,
+              "unimplemented support for the given dtype conversion");
+        }
+        Value constDtype = rewriter.create<Torch::ConstantIntOp>(
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2416]: I think it is better to have the shorter version, right?
```cpp
...<Torch::ConstantIntOp>(binder.getLoc(), rewriter.getI64IntegerAttr(dtypeIntTorch);
```

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 2]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64),
+                                    dtypeIntTorch));
+
+        SmallVector<Value> shapeValues;
+        for (auto dim : shape)
+          shapeValues.push_back(rewriter.create<Torch::ConstantIntOp>(
+              binder.getLoc(), rewriter.getI64IntegerAttr(dim)));
+
+        Value shapeList = rewriter.create<Torch::PrimListConstructOp>(
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2426]: Earlier comment about the createConstantIntList()

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 3]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+        }
+        Value constDtype = rewriter.create<Torch::ConstantIntOp>(
+            binder.getLoc(), rewriter.getType<Torch::IntType>(),
+            rewriter.getIntegerAttr(rewriter.getIntegerType(64),
+                                    dtypeIntTorch));
+
+        Value cstNone = rewriter.create<Torch::ConstantNoneOp>(binder.getLoc());
+        Value cstFalse =
+            rewriter.create<Torch::ConstantBoolOp>(binder.getLoc(), false);
+        input = rewriter.create<Torch::AtenToDtypeOp>(
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2489]: I'm a bit confused on this, and similarly with the randomNormalLike ops. If you aren't making a copy, does AtenUniformOp (or the similar op in the other conversion) make an empty memory format op at some point of the same shape as input? Or does it overwrite the contents of input and replace them with randomized values?

**Review by zjgarvey — CHANGES_REQUESTED** [review]
Thanks @vivekkhandelwal1 . 

I mostly have some questions and small changes.

A side question: Sometimes you put a lot of lit tests, and sometimes none. Do you only use lit tests when e2e-tests are insufficient?

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 4]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+            binder.op, resultType, input, cstMean, cstStd,
+            /*generator=*/cstNone);
+        return success();
+      });
+  patterns.onOp(
+      "RandomUniform", 1,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        SmallString<64> name("torch.onnx.seed");
+        auto attr = binder.op->getAttr(name);
+        if (attr) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2392]: Unneeded curly brackets.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 5]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+            binder.op, resultType, self, cstLow, cstHigh,
+            /*generator=*/cstNone);
+        return success();
+      });
+  patterns.onOp(
+      "RandomUniformLike", 1,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        SmallString<64> name("torch.onnx.seed");
+        auto attr = binder.op->getAttr(name);
+        if (attr) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2456]: Ditto

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 6]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+                                    dtypeIntTorch));
+
+        SmallVector<Value> shapeValues;
+        for (auto dim : shape)
+          shapeValues.push_back(rewriter.create<Torch::ConstantIntOp>(
+              binder.getLoc(), rewriter.getI64IntegerAttr(dim)));
+
+        Value shapeList = rewriter.create<Torch::PrimListConstructOp>(
+            binder.getLoc(),
+            Torch::ListType::get(Torch::IntType::get(binder.op->getContext())),
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2428]: `rewriter.getType<Torch::ListType>(rewriter.getType<Torch::IntType>())`

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 7]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
                       binder.op, resultType, input, cstAlpha, value);
                   return success();
                 });
+  patterns.onOp(
+      "RandomNormal", 1,
+      [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+        SmallString<64> name("torch.onnx.seed");
+        auto attr = binder.op->getAttr(name);
+        if (attr) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2273]: You should name it something more specific than `attr` in this case. E.g. `seedAttr`

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 8]
```diff
@@ -2265,4 +2265,242 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+        if (binder.tensorOperand(input) ||
+            binder.s64IntegerAttr(dtypeIntOnnx, "dtype", 1) ||
+            binder.f32FloatAttr(high, "high", 1.0) ||
+            binder.f32FloatAttr(low, "low", 0.0) ||
+            binder.tensorResultType(resultType)) {
+          return failure();
+        }
+
+        dtypeIntTorch = onnxDtypeIntToTorchDtypeInt(dtypeIntOnnx);
+        if (dtypeIntTorch == -1) {
```
Comment (rsuderman) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2476]: See comment about `std::optional`

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2273]: Yeah, the seed can be any value so the only option to check if it's present or not is this way.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 3 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2482]: The random ops during Torch->Linalg lowering creates new empty tensors in which results are stored. They don't overwrite the existing input operands. Although, the shape and dtype info is sometimes inferred from the input operand.

**Review by rsuderman — APPROVED** [review]

File: lib/Conversion/TorchOnnxToTorch/Utils.cpp [hunk 9]
```diff
@@ -59,3 +59,41 @@ bool mlir::torch::onnx_c::areAllElementsDistinct(SmallVector<int64_t> array) {
   // as array's size.
   return (set.size() == array.size());
 }
+
+std::optional<int64_t>
+mlir::torch::onnx_c::onnxDtypeIntToTorchDtypeInt(int64_t dtypeIntOnnx) {
```
Comment (zjgarvey) [inline lib/Conversion/TorchOnnxToTorch/Utils.cpp:64]: What happened to the Endian class?

**Review by zjgarvey — APPROVED** [review]
As long as Endian class isn't used anywhere and you are just cleaning it up, I think this looks good.

File: lib/Conversion/TorchOnnxToTorch/Utils.cpp (hunk 9 above)
Comment (vivekkhandelwal1) [reply lib/Conversion/TorchOnnxToTorch/Utils.cpp:64]: It's declared in Utils.h

**Pull request merged**